reggie/ingestion/preprocessor/west_virginia_preprocessor.py

```diff
@@ -1,87 +1,361 @@
+            }
+        else:
+            voter_history = voter_histories[0]
+
+            # Read voter file into pandas dataframe
+            df_history = pd.read_csv(
+                voter_history["obj"],
+                sep=config["delimiter"],
+                encoding="latin-1",
+                dtype=str,
+                header=0 if config["has_headers"] else None,
+            )
+
+            # all_history: text[]
+            # sparse_history: int[]
+            # # absentee, early-voting, regular
+            # votetype_history: text[]
+            # # Challenged (boolean)
+            # challenged_history: integer[]
+
+            # Create new column that aggregrates the type of vote
+            # TODO: This is very slow
+            df_history["votetype"] = df_history.apply(
+                lambda row: self.derive_votetype(row), axis=1
+            )
+
+            # Clean the challenged flag, which looks to only be checked
+            # if the voter voted absentee
+            df_history["fl_challenged"] = df_history["fl_challenged"].apply(
+                self.clean_challenged_value
+            )
+
+            # Group history by voter id, then attach relevante group data to
+            # voter dataframe
+            df_history_grouped = df_history.groupby("id_voter")
+            df_voters["all_history"] = df_history_grouped["id_election"].apply(list)
+            df_voters["votetype_history"] = df_history_grouped["votetype"].apply(list)
+            df_voters["challenged_history"] = df_history_grouped["fl_challenged"].apply(
+                list
+            )
+
+            # Create dataframe of valid elections
+            election_columns = ["id_election", "dt_election", "Election_Name"]
+            df_elections_all = df_history[election_columns]
+            df_elections = (
+                df_elections_all.reset_index()
+                .groupby(election_columns)
+                .count()
+                .rename({"index": "count"}, axis=1)
+                .reset_index()
+            )
+
+            # Clean up
+            df_elections["Election_Name"] = df_elections["Election_Name"].apply(
+                lambda x: x.strip()
+            )
+            df_elections = self.config.coerce_dates(
+                df_elections, col_list="hist_columns_types"
+            )
+
+            # Sort by ascending date
+            df_elections = df_elections.sort_values(by=["dt_election"]).reset_index()
+
+            # Create dict for encoding
+            array_encoding = {}
+            for k, row in df_elections.iterrows():
+                array_encoding[row["id_election"]] = {
+                    "index": k,
+                    "count": row["count"],
+                    # TODO: The documentation specifically says this should be MM/DD/YYYY,
+                    # but other files have the more standard YYYY-MM-DD
+                    # https://github.com/Voteshield/Inspector/wiki/Adding-a-State
```

Comment:
MM/DD/YYYY IS the expected format for election dates in the metadata file, in the feature generation code. But if you prefer to make these be iso format, you can make that work by adding your state to the conditionals in this code block of “format_election_dates” in features.py in Inspector:

    meta_data = meta["array_encoding"]
    if state in ["colorado", "michigan"]:
        election_date_format = "%Y-%m-%d"
    elif state == "kansas":
        election_date_format = "%Y"
    else:
        election_date_format = "%m/%d/%Y"